src/jdk.security.auth/share/classes/com/sun/security/auth/module/GssLoginModule.java

```diff
@@ -0,0 +1,656 @@
+    private static final String PWD = "javax.security.auth.login.password";
+
+    private String getWithDefault(String key, String defval) {
+        String value = (String)options.get(key);
+        return value != null ? value : defval;
+    }
+
+    /**
+     * Initialize this <code>LoginModule</code>.
+     *
+     * <p>
+     * @param subject the <code>Subject</code> to be authenticated.</p>
+     *
+     * <p>
+     * @param callbackHandler a <code>CallbackHandler</code> for
+     *                  communication with the end user (prompting for
+     *                  usernames and passwords, for example).</p>
+     *
+     * <p>
+     * @param sharedState shared <code>LoginModule</code> state.</p>
+     *
+     * <p>
+     * @param options options specified in the login
+     *                  <code>Configuration</code> for this particular
+     *                  <code>LoginModule</code>.</p>
+     */
+    // Unchecked warning from (Map<String, Object>)sharedState is safe
+    // since javax.security.auth.login.LoginContext passes a raw HashMap.
+    // Unchecked warnings from options.get(String) are safe since we are
+    // passing known keys.
+    @SuppressWarnings("unchecked")
+    public void initialize(Subject subject,
+                           CallbackHandler callbackHandler,
+                           Map<String, ?> sharedState,
+                           Map<String, ?> options) {
+
+        this.subject = subject;
+        this.callbackHandler = callbackHandler;
+        this.sharedState = (Map<String, Object>)sharedState;
+        this.options = options;
+
+        /*
+         * When sun.security.jgss.native=false (i.e., not using the system's
+         * native C/ELF/DLL GSS implementation) then there's nothing for this
+         * login module to do.  Otherwise we'd get into an infinite recursion
+         * problem due to re-entering GssLoginModule like this:
+         *
+         * Application -> LoginContext -> GssLoginModule -> Krb5 ->
+         *      GSSUtil.login -> LoginContext -> GssLoginModule -> ...
+         *
+         * It stands to reason that when sun.security.jgss.native=false the
+         * login modules corresponding to the actual GSS mechanisms coded in
+         * Java are the ones that should be acquiring their corresponding
+         * credentials.
+         *
+         * A policy like "let the application use GSS credentials but not the
+         * raw, underlying Krb5 credentials" when
+         * sun.security.jgss.native=false" could be expressible by adding a
+         * module option to Krb5LoginModule that causes it to add only GSS
+         * credentials to the Subject, not Krb5 credentials.
+         *
+         * (It has never been possible to express such a policy, so we lose
+         * nothing by punting here when sun.security.jgss.native=false.)
+         */
+        useNative = "true".equalsIgnoreCase(
+                System.getProperty("sun.security.jgss.native"));
```

Comment:
Consider `Boolean.getBoolean("sun.security.jgss.native")`

Reply:
Ok.

Reply:
FYI, this is all from having started with a copy of `Krb5LoginModule`, which must be ancient, not modern.